packages/attune/src/apt/packages_index.rs

```diff
@@ -0,0 +1,227 @@
+                size: rendered.len() as i64,
+                md5sum: hex::encode(Md5::digest(&rendered)),
+                sha1sum: hex::encode(Sha1::digest(&rendered)),
+                sha256sum: hex::encode(Sha256::digest(&rendered)),
+            },
+            packages,
+            contents: rendered,
+        }
+    }
+
+    fn render<'a>(packages: impl Iterator<Item = &'a PublishedPackage>) -> String {
+        // TODO: A better way to ensure index reproducibility is to sort the
+        // packages before we render them.
+        let mut index = packages
+            .map(|published| {
+                let pkg = &published.package;
+                let fields = pkg
+                    .paragraph
+                    .as_object()
+                    .unwrap()
+                    .into_iter()
+                    .map(|(k, v)| format!("{}: {}", k, v.as_str().unwrap()))
+                    .chain(vec![
+                        format!("Filename: {}", published.filename),
+                        format!("Size: {}", pkg.size.to_string()),
+                        format!("MD5sum: {}", pkg.md5sum),
+                        format!("SHA1: {}", pkg.sha1sum),
+                        format!("SHA256: {}", pkg.sha256sum),
+                    ])
+                    .collect::<Vec<String>>();
+                fields.join("\n")
+            })
+            .collect::<Vec<String>>()
+            .join("\n\n");
+        if index.is_empty() {
+            return String::new();
+        }
+        index.push('\n');
+        index
+    }
+
+    /// Add a package to this Packages index. This will re-render the index,
+    /// updating the size, checksums, and contents.
+    ///
+    /// If the package is already present in the index, this is a no-op.
+    pub fn add_package(&mut self, added: Package) {
+        // TODO: What if these fields are the same, but other fields (e.g. the
+        // package hashes) are different? Should we crash? Should we push that
+        // invariant checking outwards?
+        if self.packages.iter().any(|p| {
+            p.package.name == added.name
+                && p.package.version == added.version
+                && p.package.architecture == added.architecture
+        }) {
+            return;
+        }
+        self.packages
+            .push(PublishedPackage::from_package(added, &self.meta.component));
+        self.rerender();
+    }
+
+    /// Remove a package from this Packages index. This will re-render the
+    /// index, updating the size, checksums, and contents.
+    ///
+    /// If the package is not present in the index, this is a no-op.
+    pub fn remove_package(&mut self, removed: PublishedPackage) {
+        self.packages.retain(|p| {
+            !(p.package.name == removed.package.name
+                && p.package.version == removed.package.version
+                && p.package.architecture == removed.package.architecture)
+        });
+        self.rerender();
+    }
```

Comment:
[optional, since this isn't known to be a hot path]

re-rendering every time seems very wasteful, is there a reason we can't just render once after finished adding all packages?

if the reasoning is that the struct might be passed off in a partial built state, consider using the typestate pattern to prevent this: https://cliffle.com/blog/rust-typestate/

e.g. "we can only render once it has been finalized"